File: single_linked_list/go/linked_list.go
package single_linked_list

import "strconv"

type Node struct {
	item *Item
	next *Node
}

type LinkedList struct {
	head *Node
}

func (list *LinkedList) String() string {
	s := "["
	aux := list.head

	for aux != nil {
		s += strconv.Itoa(aux.item.value)
		if aux.next != nil {
			s += ", "
		}
		aux = aux.next
	}
	s += "]"
	return s
}

func CreateLinkedList() (*LinkedList, error) {
	list := &LinkedList{head: nil}
	return list, nil
}

func creatNode(item *Item) (*Node, error) {
	node := &Node{
		item: item,
		next: nil,
	}
	return node, nil
}

func (list *LinkedList) Length() int {
	var length int = 0

	aux_node := list.head
	for aux_node != nil {
		length++
		aux_node = aux_node.next
	}

	return length
}

func (list *LinkedList) AddHead(item *Item) error {
	new_node, error := creatNode(item)
	new_node.next = list.head
	list.head = new_node
	return error
}

func (list *LinkedList) AddTail(item *Item) error {
	aux_node := list.head

	if aux_node == nil {
		return list.AddHead(item)
	}

	for aux_node.next != nil {
		aux_node = aux_node.next
	}

	new_node, error := creatNode(item)
	aux_node.next = new_node
	return error
}

func add_tail_recursive(node **Node, item *Item) error {
	if *node == nil {
		new_node, error := creatNode(item)
		*node = new_node
		return error
	}
	return add_tail_recursive(&(*node).next, item)
}

func (list *LinkedList) AddTailRecursive(item *Item) error {
	return add_tail_recursive(&list.head, item)
}

func (list *LinkedList) RemoveHead() *Item {
	if list.head == nil {
		return nil
	}

	removed := list.head
	list.head = removed.next
	item := removed.item
	removed = nil
	return item
}

func (list *LinkedList) RemoveTail() *Item {
	if list.head == nil {
		return nil
	}
	var removed *Node = nil

	if list.head.next == nil {
		return list.RemoveHead()
	}

	aux := list.head
	for aux.next.next != nil {
		aux = aux.next
	}

	removed = aux.next
	item := removed.item
	aux.next = nil
	removed = nil

	return item
}

func (list *LinkedList) FindItem(item *Item) *Item {
	aux_node := list.head
	for aux_node != nil {
		if aux_node.item.value == item.value {
			return aux_node.item
		}
		aux_node = aux_node.next
	}
	return nil
}

func find_item_recursive(node **Node, item *Item) *Item {
	if *node == nil {
		return nil
	}
	if (*node).item.value == item.value {
		return (*node).item
	}
	return find_item_recursive(&(*node).next, item)
}

func (list *LinkedList) FindItemRecursive(item *Item) *Item {
	return find_item_recursive(&list.head, item)
}

File: single_linked_list/c++/linked_list.cpp
#include "linked_list.h"

#include <iostream>

using namespace std;


// Node methods
Node::Node(Item *item) {
    this->item = item;
    this->next = NULL;
}

Node::~Node() {}

Item* Node::getItem() {
    return this->item;
}

void  Node::setItem(Item* item) {
    this->item = item;
}


// Linked list methods
LinkedList::LinkedList(){
    this->head = NULL;
}


LinkedList::~LinkedList() {}

void LinkedList::print() {
    Node *aux = this->head;
    cout << "[";
    while(aux) {
        cout << aux->getItem()->getValue();
        if(aux->next) {
            cout << ", ";
        }
        aux = aux->next;
    }
    cout << "]";

}

Node* LinkedList::getHead() {
    return this->head;
}


int LinkedList::getLength() {
    Node * aux = this->head;
    int length = 0;
    while(aux) {
        length++;
        aux = aux->next;
    }
    return length;
}


void LinkedList::addHead(Item *item) {
    Node *new_node = new Node(item);
    new_node->next = this->head;
    this->head = new_node;
}

void LinkedList::addTail(Item *item) {
    Node *aux_node = this->head;

    if(!aux_node) {
        this->addHead(item);
    } else {
        while(aux_node->next) {
            aux_node = aux_node->next;
        }
        aux_node->next = new Node(item);
    }
}

static void add_tail_recursive(Node **node, Item *item) {
    if(!*node) {
        *node = new Node(item);
        return;
    }
    add_tail_recursive(&(*node)->next, item);
}

void LinkedList::addTailRecursive(Item *item) {
    add_tail_recursive(&this->head, item);
}

Item* LinkedList::removeHead() {
    if(!this->head) {
        return NULL;
    }

    Node *removed = this->head;
    this->head = removed->next;
    Item* item = removed->getItem();
    delete removed;
    return item;
}

Item* LinkedList::removeTail() {
    if(!this->head) {
        return NULL;
    }

    if(!this->head->next) {
        return this->removeHead();
    }

    Node *aux = this->head;
    while(aux->next->next) {
        aux = aux->next;
    }

    Node *removed = aux->next;
    Item *item = removed->getItem();
    aux->next = NULL;
    delete removed;

    return item;
}

Item* LinkedList::findItem(Item *item) {
    Node *aux = this->head;
    while(aux) {
        if(aux->getItem()->getValue() == item->getValue()) {
            return aux->getItem();
        }
        aux = aux->next;
    }
    return NULL;
}


static Item* find_item_recursive(Node *node, Item *item) {
    if(!node) {
        return NULL;
    }
    if(node->getItem()->getValue() == item->getValue()) {
        return node->getItem();
    }
    return find_item_recursive(node->next, item);
}

Item* LinkedList::findItemRecursive(Item *item) {
    return find_item_recursive(this->head, item);
}